src/ast/ast_enrich.c

```diff
@@ -0,0 +1,247 @@
+	if(cypher_ast_with_has_include_existing(with_clause)) {
+		const char **prev_aliases = AST_GetProjectAll(with_clause);
+		*aliases = array_ensure_append(*aliases, prev_aliases, array_len(prev_aliases), const char *);
+		return;
+	}
+
+	uint projection_count = cypher_ast_with_nprojections(with_clause);
+	for(uint i = 0; i < projection_count; i ++) {
+		const cypher_astnode_t *projection = cypher_ast_with_get_projection(with_clause, i);
+		const char *alias = NULL;
+		const cypher_astnode_t *identifier = cypher_ast_projection_get_alias(projection);
+		if(identifier == NULL) {
+			// The projection was not aliased, so the projection itself must be an identifier.
+			identifier = cypher_ast_projection_get_expression(projection);
+			assert(cypher_astnode_type(identifier) == CYPHER_AST_IDENTIFIER);
+		}
+		alias = cypher_ast_identifier_get_name(identifier);
+
+		*aliases = array_append(*aliases, alias);
+	}
+}
+
+// Sort an array of strings and remove duplicate entries.
+static void _uniqueArray(const char **arr) {
+#define MODIFIES_ISLT(a,b) (strcmp((*a),(*b)) < 0)
+	int count = array_len(arr);
+	QSORT(const char *, arr, count, MODIFIES_ISLT);
+	uint unique_idx = 0;
+	for(int i = 0; i < count - 1; i ++) {
+		if(strcmp(arr[i], arr[i + 1])) {
+			arr[unique_idx++] = arr[i];
+		}
+	}
+	arr[unique_idx++] = arr[count - 1];
+	array_trimm_len(arr, unique_idx);
+}
+
+static const char **_collect_aliases_in_scope(AST *ast, uint scope_start, uint scope_end) {
+	assert(scope_start != scope_end);
+	const char **aliases = array_new(const char *, 1);
+
+	for(uint i = scope_start; i < scope_end; i ++) {
+		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
+		cypher_astnode_type_t type = cypher_astnode_type(clause);
+		if(type == CYPHER_AST_WITH) {
+			// The WITH clause contains either aliases or its own STAR projection.
+			_collect_with_projections(clause, &aliases);
+		} else if(type == CYPHER_AST_MATCH) {
+			// The MATCH clause contains one pattern of N paths.
+			const cypher_astnode_t *pattern = cypher_ast_match_get_pattern(clause);
+			_collect_aliases_in_pattern(pattern, &aliases);
+		} else if(type == CYPHER_AST_CREATE) {
+			// The CREATE clause contains one pattern of N paths.
+			const cypher_astnode_t *pattern = cypher_ast_create_get_pattern(clause);
+			_collect_aliases_in_pattern(pattern, &aliases);
+		} else if(type == CYPHER_AST_MERGE) {
+			// The MERGE clause contains one path.
+			const cypher_astnode_t *path = cypher_ast_merge_get_pattern_path(clause);
+			_collect_aliases_in_pattern(path, &aliases);
+		} else if(type == CYPHER_AST_UNWIND) {
+			// The UNWIND clause introduces one alias.
+			const cypher_astnode_t *unwind_alias = cypher_ast_unwind_get_alias(clause);
+			aliases = array_append(aliases, cypher_ast_identifier_get_name(unwind_alias));
+		}
+	}
+
+	// Sort and unique the aliases array so that we won't make redundant projections given queries like:
+	// MATCH (a)-[]->(a) RETURN *
+	_uniqueArray(aliases);
```

Comment:
Would you prefer using rax to keep track over unique elements and build an array out of rax?

Reply:
My issue with rax here is that the current approach collects a series of query-scoped string pointers, so we don't need to worry about freeing or losing scope. rax _keys_ are scoped to the lifetime of the rax, though we could attach the same string pointers as values to the keys and then add all the values into an array.

I don't have a strong preference here, you?